### ISSUE TITLE
ASan: please add option to suppress details on indirect leaks

### ISSUE THREAD

**Issue by DDvO** [google/sanitizers#1712]
_Opening a new FR since the request to re-open [the original one](https://github.com/google/sanitizers/issues/1250) has been ignored so far.
Repeating part of the text from there:_

It would be great if it was possible to suppress the output of indirect leaks
(instead only an informational note like "Output of 1337 indirect leaks omitted." should be printed).

At OpenSSL we are having a hard time working around the limitations of `max_leaks` - 
see for instance https://github.com/openssl/openssl/issues/22303#issuecomment-1751728889

It would be much preferred to have an additional option (or maybe an extension of the existing option)
that prevents outputting (or even more efficiently, already prevents producing internally) all that messy data on indirect leaks,
while still showing all (or a limited number of) direct leaks.

Here is an example how extremely bulky and inefficient the mem leak output can be
as long as the indirect leaks are included:
https://github.com/openssl/openssl/actions/runs/6411070545/job/17405721691?pr=18916

**Comment by vitalybuka** [comment]
> while still showing all (or a limited number of) direct leaks.

I guess I am missing something. Isn't `max_leaks` current behavior is like that already?

**Comment by nhorman** [comment]
@vitalybuka not quite.  max_leaks stops reporting after the first leak, while this filter prints all direct leaks, and skips any indirect leaks resulting from a given direct leak

**Comment by vitalybuka** [comment]
> max_leaks stops reporting after the first leak,

Did you mean `max_leaks stops reporting after the first N leak` ?

Sorry if I asking already answered questions in some of these links.
Why is important to see all direct leaks, and not e.g. `max_leaks` first leaks?

**Comment by DDvO** [comment]
> Did you mean `max_leaks stops reporting after the first N leak` ?

Yes, and it does so regardless whether these leaks are direct or indirect.

> Sorry if I asking already answered questions in some of these links. Why is important to see all direct leaks, and not e.g. `max_leaks` first leaks?

Often there is a large number of indirect leaks and typically they are not interesting/helpful for detecting the cause of the leaks.
Therefore, it would be much preferred if there was an option to show direct leaks only.

**Comment by nhorman** [comment]
Suppression of indirect leaks should already occur, using the log grouping feature of GitHub ci (that is to say, direct leaks get printed to the ci log console, while indirect leaks get grouped and collapsed down into a log group you explicitly have to expand to see).

I would need to find the pr we added this in, but it must have been almost a year ago.  Has something broken with it?

**Comment by DDvO** [comment]
@nhorman do you mean your OpenSSL PR https://github.com/openssl/openssl/pull/22678?
This adds a nice workaround at OpenSSL CI level.

Yet this does not help for other uses of ASan (both within OpenSSL and with other projects),
and so what is still missing is an upstream solution.

As I wrote in the OP, it likely would be more efficient to even prevent producing the info on indirect leaks if not wanted.
In this case, it would be sufficient to give just their total number (e.g., "Output of 1337 indirect leaks omitted."), if at all.

**Comment by nhorman** [comment]
Ah, my bad.  I got this notification assuming you opened this issue in the openssl project.  I missed the fact that this was opened in the sanitizers repo.


That said, the workaround we added in openssl ci is, while not ideal, applicable in the general case.  You can record the asan output to a file and use awk to filter any output from the indirect leak string to the next blank line.

Though yes, an option like direct_only in ASAN_OPTIONS would be better.
